File: src/main/java/org/normal/common/context/ContextImpl.java
package org.normal.common.context;

import org.normal.api.context.Context;
import org.normal.api.context.ContextType;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class ContextImpl<T extends ContextType> implements Context<T> {

    private final Map<String, Object> data;
    private final T type;

    public ContextImpl(T type) {
        this.data = new HashMap<>();
        this.type = type;
    }

    @Override
    public T getType() {
        return this.type;
    }

    @Override
    public <Z> Optional<Z> getData(String key, Class<Z> type) {
        return Optional.ofNullable(data.get(key)).map(type::cast);
    }

    @Override
    public void putData(String key, Object value) {
        this.data.put(key, value);
    }
}

File: src/main/java/org/normal/common/kt/context/ContextImpl.kt
package org.normal.common.kt.context

import org.normal.api.kt.context.Context
import org.normal.api.kt.context.ContextType

class ContextImpl(override val type : ContextType) : Context {

    private val backingMap: MutableMap<String, Any> = mutableMapOf()

    override fun <data: Any> getData(key: String, type: Class<data>): data? {
        require(backingMap[key] != null) { "Data does not exist" }

        return backingMap[key] as data?
    }

    override fun <data : Any> putData(key: String, value: data) {
        backingMap[key] = value
    }



    fun getBackingMap(): Map<String, Any> {
        return backingMap.toMap()
    }
}